### DISCUSSION TITLE
Should we merge all the factorial algorithms together?

### DISCUSSION THREAD

**Discussion by realstealthninja** [TheAlgorithms/C-Plus-Plus#2454] [Ideas]
The factorial algorithms are either similar or just various implementations of a single algorithm
now this is not an issue, but it is a bit disorganized.
There are pros and cons to doing such a thing.

pros:
 - organized repository
 - easier maintenance

cons:
 - users *may* be confused about the changes
 - new implementations will have to be directed their specific files
 - would have to delete old files

### factorial algorithms
  - [check factorial](https://github.com/TheAlgorithms/C-Plus-Plus/blob/4b740d464c8e8898162181b615efe7884d1980be/math/check_factorial.cpp)
  - [factorial](https://github.com/TheAlgorithms/C-Plus-Plus/blob/4b740d464c8e8898162181b615efe7884d1980be/math/factorial.cpp)
  - [double factorial](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/double_factorial.cpp)
  - [large factorial](https://github.com/TheAlgorithms/C-Plus-Plus/blob/4b740d464c8e8898162181b615efe7884d1980be/math/large_number.h)
  - #2453

There might be more examples of algorithms in disarray lurking in unexplored directories

**Comment by appgurueu** [comment]
Checking whether a number is a factorial is not quite the same as computing factorials (though it could be implemented in terms of computing factorials). Some implementations also seem to differ in whether they are recursive/iterative.

I suppose this should be consolidated as follows:

* Generic check factorial method (any type which provides the necessary arithmetic operations should work).
* Generic recursive and iterative factorial methods. IMO the recursive variant could be eliminated as strictly inferior (or it should at least get a warning).

**Reply by realstealthninja** [reply]
we do have some checking built into some algorithm files i cant remember them off the top of my head but i think there are some in the repository

i mean mentioning that something strictly inferior would have us enforcing a big O performance rating on each which doesnt sound that bad

**Comment by sorohere** [comment]
I get the points you made. After considering @appgurueu  thoughts, here's a quick solution:

Separate methods for factorial checking and calculation.
Create a general "check factorial" method for all number types.
Keep both recursive and iterative factorial methods, but highlight the better one.
New implementations in separate files for clarity.
Decide on existing checks – separate methods or integrate?

Adding big O performance rating is a good call.
